types/decimal.go

```diff
@@ -0,0 +1,324 @@
+func NewDec(i, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(big.NewInt(i), precisionMultiplier(prec)),
+	}
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec !> Precision
+func NewDecFromBigInt(i *big.Int, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(i, precisionMultiplier(prec)),
+	}
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec !> Precision
+func NewDecFromInt(i Int, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(i.BigInt(), precisionMultiplier(prec)),
+	}
+}
+
+// create a decimal from a decimal string (ex. "1234.5678")
+func NewDecFromStr(str string) (d Dec, err Error) {
+	if len(str) == 0 {
+		return d, ErrUnknownRequest("decimal string is empty")
+	}
+
+	// first extract any negative symbol
+	neg := false
+	if string(str[0]) == "-" {
+		neg = true
+		str = str[1:]
+	}
+
+	if len(str) == 0 {
+		return d, ErrUnknownRequest("decimal string is empty")
+	}
+
+	strs := strings.Split(str, ".")
+	lenDecs := 0
+	combinedStr := strs[0]
+	if len(strs) == 2 {
+		lenDecs = len(strs[1])
+		if lenDecs == 0 || len(combinedStr) == 0 {
+			return d, ErrUnknownRequest("bad decimal length")
+		}
+		combinedStr = combinedStr + strs[1]
+	} else if len(strs) > 2 {
+		return d, ErrUnknownRequest("too many periods to be a decimal string")
+	}
+
+	if lenDecs > Precision {
+		return d, ErrUnknownRequest("too much Precision in decimal")
+	}
+
+	// add some extra zero's to correct to the Precision factor
+	zerosToAdd := Precision - lenDecs
+	zeros := fmt.Sprintf(`%0`+strconv.Itoa(zerosToAdd)+`s`, "")
+	combinedStr = combinedStr + zeros
+
+	combined, ok := new(big.Int).SetString(combinedStr, 10)
+	if !ok {
+		return d, ErrUnknownRequest("bad string to integer conversion")
+	}
+	if neg {
+		combined = new(big.Int).Neg(combined)
+	}
+	return Dec{combined}, nil
+}
+
+//nolint
+func (d Dec) IsZero() bool      { return (d.Int).Sign() == 0 } // Is equal to zero
+func (d Dec) Equal(d2 Dec) bool { return (d.Int).Cmp(d2.Int) == 0 }
+func (d Dec) GT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == 1 }             // greater than
+func (d Dec) GTE(d2 Dec) bool   { return !d.LT(d2) }                            // greater than or equal
+func (d Dec) LT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == -1 }            // less than
+func (d Dec) LTE(d2 Dec) bool   { return !d.GT(d2) }                            // less than or equal
+func (d Dec) Neg() Dec          { return Dec{new(big.Int).Neg(d.Int)} }         // Is equal to zero
+func (d Dec) Add(d2 Dec) Dec    { return Dec{new(big.Int).Add(d.Int, d2.Int)} } // addition
+func (d Dec) Sub(d2 Dec) Dec    { return Dec{new(big.Int).Sub(d.Int, d2.Int)} } // subtraction
+
+// multiplication
+func (d Dec) Mul(d2 Dec) Dec {
+	mul := new(big.Int).Mul(d.Int, d2.Int)
+	chopped := BankerRoundChop(mul, Precision)
+	return Dec{chopped}
+}
+
+// quotient
+func (d Dec) Quo(d2 Dec) Dec {
+	mul := new(big.Int).Mul(new(big.Int).Mul( // multiple Precision twice
```

Comment:
We shouldn't create two seperate big ints here.
Instead this would be better as:
```
mul :=new(big.Int).Mul(d.Int, precisionInt())
mul.Mul(mul, d.precisionInt())
```

Reply:
added